### PR TITLE
Add Rails generators

### PR DESCRIPTION
**Pull request by gee-forr**
TL;DR - LightService is now bundled with generators for use within Rails.

As part of my drive to ensure LightService is adopted as much as possible with the people I work with, I thought providing rails generators would be a really useful way to help developers get traction with building software using LS.

This is a rough work-in-progress for now, but I thought it would be useful to get the code in early, and get any feedback, guidance, or help, before I go too far in a potentially wrong direction.

@adomokos - thanks again for this wonderful gift we rely on everyday. I hope you like the PR and find it a good fit for the project.

### Todo

- [x]  Update readme with generator documentation
- [x]  Ensure that generators work when LS is used as a gem (have only built them direct in rails right now, and they probably won't work)
- [X]  Generate actions
- [X]  Generate rspec specs for actions, unless `--no-tests`
- [x] Generated tests should all be set to "pending" with `xit`
- [x]  Add usage files for action generator
- [X]  Support module namespaced actions
- [X]  Support custom actions dir (defaults to `/app/actions`)
- [X]  Support defining expects and promises on the CLI
- [x]  Generate organizers
- [x]  Generate rspec specs for organizers, unless `--no-tests`
- [x]  Add usage files for organizer generator
- [x]  Support module namespaced organizers
- [x]  Support custom organizer dir (defaults to `/app/organizers`)
- [x]  DRY up code into modules
- [x]  Functional tests for generated actions
- [x]  Functional tests for generated organizers
- [ ]  ~~Generated tests for actions can execute via rspec~~ - not viable
- [ ]  ~~Generated tests for organizers can execute via rspec~~ - not viable
- [x] Address all rubocop violations

### CODE REVIEW

**Comment by gee-forr** [comment]
This is the command help so far for actions:

```shell
$ rails g light_service:action
Running via Spring preloader in process 59103
Usage:
  rails generate light_service:action NAME [expects:one,thing promises:something,else] [options]

Options:
  [--skip-namespace], [--no-skip-namespace]  # Skip namespace (affects only isolated applications)
  [--dir=DIR]                                # Path to write actions to
                                             # Default: actions
  [--tests], [--no-tests]                    # Generate test files (currently only RSpec supported)
                                             # Default: true
  [--roll-back], [--no-roll-back]            # Add a roll back block
                                             # Default: true
```

**Comment by gee-forr** [comment]
Here's how to generate an action:
```shell
rails g light_service:action Foo::Bar expects:one,two promises:three,four

Running via Spring preloader in process 60248
      create  app/actions/foo/bar.rb
      create  spec/actions/foo/bar_spec.rb
```

And here's the corresponding files:

```ruby
# frozen_string_literal: true

module Foo
  class Bar
    extend ::LightService::Action

    expects  :one, :two
    promises :three, :four

    executed do |ctx|
      one = ctx.one
      two = ctx.two
    end

    rolled_back do |ctx|
    end
  end
end
```

and 

```ruby
# frozen_string_literal: true

require 'rails_helper'

RSpec.describe Foo::Bar, type: :action do
  subject { described_class.execute(ctx) }

  let(:ctx) do
    {
      one: nil,
      two: nil,
    }
  end

  context "when executed" do
    it "is expected to be successful" do
      expect(subject).to be_a_success
    end

    it "is expected to promise 'three'" do
      expect(subject.three).to eq SomeThreeClass
    end

    it "is expected to promise 'four'" do
      expect(subject.four).to eq SomeFourClass
    end
  end
end
```

**Comment by adomokos** [comment]
This is a fantastic idea! 😍 

I have to admit, I have not used LS with Rails for 4-5 years now, I always used it in Rails-independent Ruby command-line apps. However, this could be very useful for teams using LS with Rails.

Please go ahead with this change, I'll definitely merge it when it's in good shape.

One thing that would be very important to me is adding functional tests that runs the generators, creates the files in `/tmp` or something similar, and runs the generated tests as well as part of the build process.

**Comment by dominicsanto** [comment]
I think this change will be a great addition and give it that "Railsy" feel. Looking forward to having this in my next version when i update

**Comment by adomokos** [comment]
Thank you for your hard work on this @gee-forr, let me know when I can start taking a look at this.

**Comment by gee-forr** [comment]
Hey @adomokos - I think is this ready for you to take a look at.

I unfortunately was not able to add running the generated specs as part of the test suite, as that would involve pulling in too many rails components, and would make the gem extremely heavy for very little extra added value.

I noticed that there are some tests failing on CI that are not related to any code changes in this particular PR, so I'm not sure how you want to handle them?

Looking forward to the review, and hopefully making this a part of LS.

**Comment by adomokos** [comment]
It seems the ActiveSupport 3.0 checks failed. I removed those appraisals, see [this](https://github.com/adomokos/light-service/pull/197) PR.
Can you please rebase from `master`? Let's get to a 🟢  build before we look at this PR in detail. Thanks!

**Comment by gee-forr** [comment]
whoop whoop! all green. over to you @adomokos :)

File: README.md [hunk 1]
```diff
@@ -857,9 +858,57 @@ This context then can be passed to the action under test, freeing you up from th
...
+
+### Advanced action generation
+
+You are able to optionally specify `expects` and/or `promises` keys during generation
+
+```shell
+rails generate light_service:action CrankWidget expects:one_fish,two_fish promises:red_fish,blue_fish
+```
+
+When specifying `expects`, convenience variables will be initialised in the `executed` block so that you don't have to call
```
Comment (adomokos) [inline README.md:903]: ~initialised~ - typo, I'd use `initialized`

**Comment by adomokos** [comment]
@gee-forr - started taking a look at it. I am on vacation atm, it might take a little while to get it all reviewed.

Thank you for all your hard work on this, I'll definitely merge this PR!!

File: README.md (hunk 1 above)
Comment (gee-forr) [reply README.md]: My bad - we use UK English in South Africa, this one slipped through the cracks. It's fixed now.

**Comment by gee-forr** [comment]
> @gee-forr - started taking a look at it. I am on vacation atm, it might take a little while to get it all reviewed.
> 
> Thank you for all your hard work on this, I'll definitely merge this PR!!

No worries man - take your time and enjoy your vacation. The real world will hopefully still be here when you get back. No guarantees, this is 2020, after all.

**Comment by adomokos** [comment]
Hello @gee-forr, I am back again. Will look at this PR by end of this week. Cheers!

File: spec/lib/generators/full_generator_test_blobs.rb [hunk 2]
```diff
@@ -0,0 +1,193 @@
+# rubocop:disable Metrics/ClassLength, Metrics/MethodLength
+class FullGeneratorTestBlobs
```
Comment (puffo) [inline spec/lib/generators/full_generator_test_blobs.rb:2]: Perhaps namespace this pseudo-utility/pseudo-fixture under `/spec/lib/generators/utils/` (or something like `/spec/lib/utilities` ) to increase clarity of intent, and also allow a future option to split up into smaller files if this gets unwieldy.

File: spec/lib/generators/full_generator_test_blobs.rb [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      "\n" \
+      "    executed do |ctx|\n" \
+      "      foo = ctx.foo\n" \
+      "      bar = ctx.bar\n" \
+      "    end\n" \
+      "  end\n" \
+      "end"
+  end
+
+  def self.advanced_action_spec_blob
```
Comment (puffo) [inline spec/lib/generators/full_generator_test_blobs.rb:62]: This general strategy is a great way to ensure that the generators are  pumping out the expected static files, but they are slightly brittle when trying to test the dynamic arguments available to the generator.

Would it be at all useful to supply some of those dynamic input values, like `expects_inputs: [:foo, :bar]` and `promises_inputs: [:baz, :qux]` as arguments to this method, and then interpolate those values in the returned blob?

That way tests might better test the dynamic behaviour of the generators.

For example, if something changes in `lib/generators/light_service/templates/action_template.erb`, over here:
```
<%= indent %>expects  <%= @expects.map { |k| ":#{k}" }.join(', ') %>
``` 
having those `expects_inputs` returned by the blob could help detect any regressions if those expressions are inadvertently changed.

Comment (puffo) [reply spec/lib/generators/full_generator_test_blobs.rb:62]: The additional coverage might not be worth the effort, though.

File: lib/generators/light_service/action_generator.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+          Will create the boilerplate for an action. Pass it an action name, e.g.
+            thing_doer, or ThingDoer   - will create ThingDoer in app/actions/thing_doer.rb
+            thing/doer, or Thing::Doer - will create Thing::Doer in app/actions/thing/doer.rb
+
+        Expects & Promises:
+          Specify a list of expected context keys by passing expects and a comma separated
+          list of keys. Adds keys to the `expects` list, creates convenience variables in
+          the action, and generates a stub context in generated specs.
+
+            expects:foo,bar,baz
```
[the remark this answers is not in the document]
Comment (adomokos) [reply lib/generators/light_service/action_generator.rb:42]: @puffo: What is wrong with `foo,bar,baz`?

File: lib/generators/light_service/organizer_generator.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                   :default => true,
+                   :desc => "Generate tests (currently only RSpec supported)"
+
+      source_root File.expand_path('templates', __dir__)
+
+      desc <<~DESCRIPTION
+        Description:
+          Will create the boilerplate for an organizer. Pass it an organizer name, e.g.
+            thing_maker, or ThingMaker   - will create ThingMaker in app/organizers/thing_maker.rb
+            thing/maker, or Thing::Maker - will create Thing::Maker in app/organizers/thing/maker.rb
```
[the remark this answers is not in the document]
Comment (adomokos) [reply lib/generators/light_service/organizer_generator.rb:26]: I like the `FooBar` suggestion. Can we switch to that?

File: lib/generators/light_service/organizer_generator.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        Options:
+          Skip rspec test creation with --no-tests
+          Write organizers to a specified dir with --dir="workflows". Default is "organizers" in app/organizers
+
+        Full Example:
+          rails g light_service:organizer My::Awesome::Organizer
+      DESCRIPTION
+
+      # rubocop:disable Metrics/MethodLength,Metrics/AbcSize
+      def create_organzier
```
[the remark this answers is not in the document]
Comment (adomokos) [reply lib/generators/light_service/organizer_generator.rb:37]: Good catch! 👍

**Comment by adomokos** [comment]
I've tried this latest version with Rails, I dig the output I see when I listed the available generators:

```
% be rails generate
Usage: rails generate GENERATOR [args] [options]

General options:
  -h, [--help]     # Print generator's options and usage
  -p, [--pretend]  # Run but do not make any changes
  -f, [--force]    # Overwrite files that already exist
  -s, [--skip]     # Skip files that already exist
  -q, [--quiet]    # Suppress status output

Please choose a generator below.

Rails:
  application_record
  assets
  channel
  controller
  generator
  helper
  integration_test
  jbuilder
  job
  mailbox
  mailer
  migration
  model
  resource
  scaffold
  scaffold_controller
  system_test
  task

ActiveRecord:
  active_record:application_record

LightService:
  light_service:action
  light_service:organizer

TestUnit:
  test_unit:channel
  test_unit:generator
  test_unit:mailbox
  test_unit:plugin
```

**Comment by adomokos** [comment]
Ran the generator like this:

`be rails g light_service:action FooBar expects:var1,var2 promises:var3,var4 -f`

It generated the files in the correct directories with the correct file/class names.

However, this is the action it generated:

```
# frozen_string_literal: true

class FooBar
  extend ::LightService::Action

  expects  :var1, :var2
  promises :var3, :var4

  executed do |ctx|
    var1 = ctx.var1
    var2 = ctx.var2
  end

  rolled_back do |ctx|
  end
end
```

In the executed block, why not:

```
  executed do |ctx|
    var3 = ctx.var1
    var4 = ctx.var2
  end
```

That way in the spec we don't need to comment out the specs:

```
    it "is expected to promise 'var3'" do
      expect(subject.var3).to eq v1
    end

    it "is expected to promise 'var4'" do
      expect(subject.var4).to eq v2
    end
```

This might be an event better starting point for newcomers. What do you think?

**Comment by adomokos** [comment]
The organizer was generated properly when I generated one from the test Rails project. 👍

File: lib/generators/light_service/action_generator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adomokos) [reply lib/generators/light_service/action_generator.rb:42]: Ha, ended up using `var1`, `var2`, `var3` in my own example... I guess this makes sense as we have more than 2 variables to go with.

File: README.md [hunk 7]
```diff
@@ -175,7 +181,143 @@ end
...
+
+if outcome.success?
+  puts outcome.greeting # which was a promised context value
+else # outcome.failure? would == true
+  puts "Rats... I can't say hello to you"
+end
+```
+
+You will notice that actions are set up to promote simplicity, i.e. they either succeed or fail, and they have
+very clear inputs and outputs. Ideally, they should do exactly one thing. This makes them as easy to test as unit tests.
```
Comment (adomokos) [inline README.md:244]:
```suggestion
very clear inputs and outputs. Ideally, they should do [exactly one thing](https://en.wikipedia.org/wiki/Single-responsibility_principle). This makes them as easy to test as unit tests.
```

File: README.md [hunk 8]
```diff
@@ -175,7 +181,143 @@ end
...
+else # outcome.failure? is true, and we can pull the failure message out of the context for feedback to the user.
+  puts "Rats... I can't say hello to you, because #{outcome.message}."
+end
+```
+
+Because organizers generally run through complex business logic, and every action has the potential to cause a failure,
+testing an organizer is functionally equivalent to an integration test.
+
+For further examples, please visit the project's [Wiki](https://github.com/adomokos/light-service/wiki) and review
+the ["Why LightService" section](#why-lightservice) above.
```
Comment (adomokos) [inline README.md:320]: This is fantastic content. I wonder if it should be added to the wiki and directly linked from here? I am OK to keep it if you think it serves its purpose best here.

File: lib/generators/light_service/generator_utils.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
+module LightService
+  module Generators
+    module GeneratorUtils
+      def make_nested_dir(dir)
+        FileUtils.mkdir_p(dir)
+      end
+
+      def supported_test_frameworks
+        %i[rspec]
+      end
```
Comment (adomokos) [inline lib/generators/light_service/generator_utils.rb:10]: Can we check what test framework is set up with the Rails project when the generator is executed? If there's no RSpec, maybe we could skip the test generator task as is.

File: spec/lib/generators/action_generator_advanced_spec.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+      my/fancy/action
+      expects:foo,bar
+      promises:baz,qux
+      --no-roll-back
+      --dir=services
+    ]
+
+    specify do
+      expect(destination_root).to(have_structure do
+        directory "app/services/my/fancy" do
```
Comment (adomokos) [inline spec/lib/generators/action_generator_advanced_spec.rb:29]: I looked at the directory structure after I ran the tests and I noticed that these tests are "polluting" the code's directory structure. 

Take a look at this:

```
% tree app
app
├── actions
├── organizers
├── processes
│   └── my
│       └── fancy
└── services
    └── my
        └── fancy

8 directories, 0 files
```

And this:

```
% tree spec/services
spec/services
└── my
    └── fancy

2 directories, 0 files
```

I am not happy with the tests creating these directories. Can you use the host OS' `/tmp` dir for this purpose?

File: lib/generators/light_service/generator_utils.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        # the selected test framework, and switch templates accordingly
+        :rspec
+      end
+
+      def must_gen_tests?
+        options.tests? && test_framework_supported?
+      end
+
+      # rubocop:disable Metrics/AbcSize
+      def create_required_gen_vals_from(name)
```
Comment (adomokos) [inline lib/generators/light_service/generator_utils.rb:29]: I'd shy away from abbreviations. Consider renaming this method to `create_required_generator_values_from()`.

File: lib/generators/light_service/organizer_generator.rb [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        gen_vals = create_required_gen_vals_from(name)
```
Comment (adomokos) [inline lib/generators/light_service/organizer_generator.rb:38]: Consider renaming this to `generated_values`.

File: lib/generators/light_service/organizer_generator.rb [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+        return unless must_gen_tests?
+
+        spec_dir       = File.join('spec', root_dir, *file_path)
+        spec_file_name = gen_vals[:spec_file_name]
+        spec_file      = "#{spec_dir}/#{spec_file_name}"
+
+        make_nested_dir(spec_dir)
+        template("organizer_spec_template.erb", spec_file)
+      end
```
Comment (adomokos) [inline lib/generators/light_service/organizer_generator.rb:62]: Might be too big of a refactor or even an overkill, but I see an organizer and a couple of actions here. 😉

Comment (adomokos) [reply lib/generators/light_service/organizer_generator.rb:62]: Ha, I looked at the `ActionGenerator` and `OrganizerGenerator` in a split view, it's almost the same code. I see two actions here with an organizer that is reusable between these two generators.

create-action-and-organizer

Comment (adomokos) [reply lib/generators/light_service/organizer_generator.rb:62]: You already worked on this PR a ton, I can refactor it after I merged it. No big deal if you don't want to do it.

Comment (gee-forr) [reply lib/generators/light_service/organizer_generator.rb:62]: Naah - it's cool, I'll refactor it. It was even more duplicated before I put `create_required_gen_vals` function in. Should be a simple enough effort - will handle it today :) Thanks for the feedback.

File: lib/generators/light_service/action_generator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (gee-forr) [reply lib/generators/light_service/action_generator.rb:42]: I'll make this change :)

File: lib/generators/light_service/action_generator.rb [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+          Specify promised context keys in the same manner as 'expects' above. This adds
+          keys to the `promises` list, and creates stub expectations in generated specs.
+
+            promises:quux,quark
```
[the remark this answers is not in the document]
Comment (gee-forr) [reply lib/generators/light_service/action_generator.rb:47]: same here.

File: lib/generators/light_service/action_generator.rb [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+          Skip ActionRollback creation with --no-roll-back
+          Write actions to a specified dir with --dir="services". Default is "actions" in app/actions
+
+        Full Example:
+          rails g light_service:action My::Awesome::Action expects:foo,bar promises:baz,qux
+      DESCRIPTION
+
+      # rubocop:disable Metrics/MethodLength,Metrics/AbcSize
+      def create_action
+        gen_vals = create_required_gen_vals_from(name)
```
[the remark this answers is not in the document]
Comment (gee-forr) [reply lib/generators/light_service/action_generator.rb:60]: Yeah  - not a great name. Will resolve.

File: lib/generators/light_service/organizer_generator.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (gee-forr) [reply lib/generators/light_service/organizer_generator.rb:26]: I wanted to have a naming distinction between actions and organizers, that made it clear they are meant to work together, which is why I chose thing::doer and thing::maker.

No sweat though, will update.

File: spec/lib/generators/full_generator_test_blobs.rb (hunk 2 above)
Comment (gee-forr) [reply spec/lib/generators/full_generator_test_blobs.rb:2]: Agreed... will make the change :)

**Comment by gee-forr** [comment]
> Ran the generator like this:
> 
> `be rails g light_service:action FooBar expects:var1,var2 promises:var3,var4 -f`
>
>   executed do |ctx|
>     var1 = ctx.var1
>     var2 = ctx.var2
>   end
>
> In the executed block, why not:
> 
> ```
>   executed do |ctx|
>     var3 = ctx.var1
>     var4 = ctx.var2
>   end
> ```
> 
> This might be an event better starting point for newcomers. What do you think?

In my experience there's rarely a 1:1 correlation between the number of expects and promises, so I didn't want to make an assumption about what the expects result in returning.

I thought about generating the tests as actual running tests, but the tests on their own are pretty low-value, in that they aren't really testing the nature of the output, just their presence.

Finally, I took a leaf out of the RSpec generator's book, and made the tests pending by default like they do. Perhaps what I should do is activate the standard "it succeeds" test, and put in proper `pending("blah")` messages for the rest?

File: README.md (hunk 8 above)
Comment (gee-forr) [reply README.md:320]: I wanted to make it easy for beginners to jump on board. Most readmes have a quick getting started section which is what people are scanning for when they want instruction. I personally love the "Why LightService" section, but I worry that new people need a quicker simpler onboarding.

I have some ideas for advanced techniques that should go into the wiki though :)

File: lib/generators/light_service/generator_utils.rb (hunk 9 above)
Comment (gee-forr) [reply lib/generators/light_service/generator_utils.rb:10]: In order to do this, we need more components added into the gem from the rails ecosystem.

I was detecting the framework before, and bailing if it wasn't `rspec` but the minute I ran the tests in isolation, they failed because `Rails.application` was not available.

This was the code I was using to find the test framework which I removed.

```ruby
      def supported_test_frameworks
        %i[rspec]
      end

      def test_framework_supported?
        supported_test_frameworks.include? test_framework
      end

      def test_framework
        # Don't know a better way to get to this value, unfortunately.
        Rails.application.config.generators.options[:rails][:test_framework]
      end

      def must_gen_tests?
        options.tests? && test_framework_supported?
      end
```

It's possible that `Rails.application.config.generators.options[:rails][:test_framework]` is not the right way to find the test framework. I'll reach out and see how it's done.

File: spec/lib/generators/action_generator_advanced_spec.rb (hunk 10 above)
Comment (gee-forr) [reply spec/lib/generators/action_generator_advanced_spec.rb:29]: Hmmm... good point. I can totally try, it might be an option in the `generator_spec` library I was using. I'll dig into its code and see if this is possible.

I do clean up after each test run though with an `after` block.

**Comment by gee-forr** [comment]
I've made a lot of the changes requested in the review. @puffo and I will do some pairing this week on getting the generators to use lightservice organizers and actions, and then we should be done...

**Comment by adomokos** [comment]
> Finally, I took a leaf out of the RSpec generator's book, and made the tests pending by default like they do. Perhaps what I should do is activate the standard "it succeeds" test, and put in proper pending("blah") messages for the rest?

I like that. 👍  I'll merge this PR once your refactoring is complete.

**Comment by adomokos** [comment]
Hey @gee-forr, are you still working on this PR? I can merge it if you don't have time (or tired of it) to make the changes. Thanks!

**Comment by gee-forr** [comment]
> Hey @gee-forr, are you still working on this PR? I can merge it if you don't have time (or tired of it) to make the changes. Thanks!

Hi @adomokos - yeah - I think it's best if you merge it in now.

I'm not tired of the PR, I'm just currently under an abnormal load of work right now, and cannot dedicate any time to OSS.

I'll come back to the generator code in another PR when I've got a little more free time and refactor it, but I don't think we should hold back this useful feature because my personal todo list is miles long.

**Comment by adomokos** [comment]
Thanks @gee-forr for adding this!

**Comment by gee-forr** [comment]
Yay! Looking forward to the rubygems release :)

**Comment by adomokos** [comment]
Was thinking about doing a bit of refactoring, but I think I'll just release it as is tonight and we'll go from there.
